design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,156 @@
+parameters:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+The resource claim below specifies that it is of `kind: MySQLInstance`, and references the above resource class in its `classRef`.
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  classRef:
+    name: standard-mysql
+    namespace: crossplane-system
+  engineVersion: "9.6"
+```
+
+If using a default resource class, the class and claim would be adjusted as follows.
+
+Resource Class:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+  labels:
+    mysqlinstance.storage.crossplane.io/default: "true"
+parameters:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+Resource Claim:
+
+```yaml
+apiVersion: storage.crossplane.io/v1alpha1
+kind: MySQLInstance
+metadata:
+  name: mysql-claim
+  namespace: demo
+spec:
+  engineVersion: "9.6"
+```
+
+This model is powerful because it allows an application developer to create a resource claim without having to know the implementation details or even the underlying provider. However, the fact that every resource class if of the same `kind` presents a key issue: The required parameters for a resource class may vary widely, and they are currently only provided as an arbitrary map that is eventually read by the controller for the the specified `provisioner`. Therefore, an administrator who is creating resource classes does not know what fields are required and will not be notified of missing or extraneous fields until the provisioning of a resource that references the class.
+
+This problem can be solved by the implementation of "strongly typed" resource classes. Instead of every resource class being of `kind: ResourceClass`, they would map one-to-one with the concrete managed resource kind for which they provided specifications (i.e. `kind: RDSInstanceClass`).
+
+Unfortunately, this makes it much more difficult to identify default resource classes because the default class controller currently watches for objects of `kind: ResourceClass`. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller cannot determine all class kinds to watch.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+## Proposal
```

Comment:
I wonder if we could do what `Defaulter` is doing by using a mutating webhook instead of a new API type. 🤔

https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.15/#mutatingwebhook-v1beta1-admissionregistration-k8s-io

Per the above (scant) documentation, Kubernetes recommends using mutating admission webhooks to set defaults for custom resources. Mutating webhooks are configured by creating a `MutatingWebhookConfiguration` resource in the Kubernetes API that informs Kubernetes to process operations on particular resource kinds before they're committed to the API server.

Given the above, perhaps each cloud provider stack could expose a mutating webhook for each `claim:class` tuple it was aware of. Each webhook would, if presented with a claim with an empty `classRef`, list all of the resource classes they're concerned with using the same label style we use today, for example `mysqlinstance.database.crossplane.io/default: true`. If exactly one default existed, it would be set. Note that nothing would actually be sent to these webhooks until a `MutatingWebhookConfiguration` requested such behaviour.

It would be the responsibility of each infrastructure stack to add a `MutatingWebhookConfiguration` for each ` claim:class` tuple with which it is concerned. Presume each infrastructure stack contained a controller that tried to create the necessary `MutatingWebhookConfigurations`, and labelled these defaulting `MutatingWebhookConfigurations` with something like `mysqlinstance.database.crossplane.io/defaulter: true`. This controller could simply check whether any other `MutatingWebhookConfiguration` already existed with the label `mysqlinstance.database.crossplane.io/defaulter: true` and refuse to add a new, conflicting `MutatingWebhookConfiguration` that would compete to set the default resource class for that claim kind.

Reply:
I had thought about this pattern initially, mostly because I wrote some [documentation](https://github.com/kubernetes-sigs/controller-runtime/pull/527/files?short_path=1c0f522#diff-1c0f522a61adc59307209c8e0296db49) this morning for the `controller-runtime` project examples that implement mutating webhooks. The drawback I saw to this pattern was that it seems to place more responsibility on the controller developer whereas having the `Defaulter` custom resource allows for a single default class controller per claim (similar to how it is currently implemented). It looks like you have a strategy for having all of these potential mutating webhooks work in concert with each other, but I may have to get some further explanation for me to wrap my head around your proposal here.

Reply:
Did some further research here and think I have a better understanding of what you are suggesting. I think that we could just have a single controller (in Crossplane core) for each claim kind (kind of how we do now) for watching creation of `MutatingWebhookConfigurations`. Essentially, the current default controllers would be ensuring the existence of only one `MutatingWebhookConfiguration` for the claim kind rather that ensuring only one default class for the claim kind (as they do currently). Then, the single mutating webhook that has been registered for a claim kind would be responsible for setting the default class for a claim by seeing if a default class exists for its strongly typed resource class kind.

The main drawbacks I see from this pattern are as follows:
- In order to have the ability to create a strongly typed default resource class, you first have to register that resource class's defaulting mutating webhook, then create the default resource class itself. Registering of mutating webhooks seems to require a bit more expertise than we may want to require for for just being able to achieve defaulting behavior.
- If a mutating webook is deregistered, but a resource class of its kind labeled as default still exists, it will look like it is serving as default but actually will not be. There could be a situation where there are many kinds of resource classes denoted as default (i.e. an `RDSInstanceClass`, `CloudSQLInstanceClass`, and `AzureMySQLInstanceClass` all having label `mysqlinstance.database.crossplane.io/defaulter: true`, and you would have to check the registered mutating webhook (i.e. `RDSMySQLDefaulter`) to see which would actually be used. It makes me somewhat uncomfortable to be able to label a resource class as default without it actually serving as the default.
- I like the ability to consistently declare a resource class as default in the same manner. If we use mutating webhooks, we lose the standardization around how defaulting works and leave the implementation up to the creation of each infrastructure stack. For example, what if the GCP infrastructure stack had a defaulting mutating webhook for MySQL to CloudSQL that said something like you are allowed to define 3 CloudSQL resource classes as default and the mutating webhook will select the most recently created one when defaulting a claim for `MySQLInstance`. Then, the AWS infrastructure stack defined a similar webhook for RDS that said you can have 7 RDS resource classes as default and the mutating webhook will select the oldest created one when defaulting a claim for `MySQLInstance`. This is a bit of a convoluted example and we could certainly provide plenty of boilerplate code to help stack developers adhere to a standard pattern, but there is still the possibility they do not.
- In the current way that we specify default resource classes and in the way proposed with mutating webhooks, the defaulting behavior is tightly coupled to the resource class object. That means that if we want to change a default we have to modify / delete the resource class. On the other hand, if using the `Defaulter` pattern proposed, the creation of a defaulting behavior is separate from the creation of resource classes. Imagine if we have 3 resources classes for RDS MySQL: `small`, `medium`, and `large`. If for a time we want `large` to serve as the default for `MySQLInstance` claims, we simply create a `Defaulter` that references `large`. Then, if we want to change the default to be `small`, we simply delete that existing `Defaulter` and create a new one that references `small`. The point here is that the lifecycle of the resource classes is completely independent of defaulting behavior, which feels like a good pattern from an end user perspective.

I firmly believe that both of these patterns are valid and could work, but wanted to make my thoughts on why the `Defaulter` pattern feels preferable to me publicly visible for consumption and discussion. Let me know if the listed points resonate with you or if any of my assumptions are off-base. Thanks for your feedback :)

Reply:
For community reference, here are two example of mutating webhook implementations:

https://github.com/programming-kubernetes/pizza-crd/blob/master/pkg/webhook/admission/pizzamutation.go

https://github.com/kubernetes-sigs/controller-runtime/blob/72ab3feaa306ee16af7aa2bd9a6b916b042bffd3/examples/crd/pkg/resource.go#L106

Reply:
We discussed this out of band today and I'm in agreement that the webhook approach doesn't have a great UX compared to a `Defaulter`. I'll resolve this comment thread.